linux/devdoc/file_linux_requirements.md

```diff
@@ -0,0 +1,159 @@
+
+**SRS_FILE_LINUX_43_033: [** If `user_callback` is `NULL` then `file_write_async` shall fail and return `FILE_WRITE_ASYNC_INVALID_ARGS`. **]**
+
+**SRS_FILE_LINUX_43_019: [** `file_write_async` shall create a `FILE_LINUX_IO` struct with `handle` as `handle`, `FILE_ASYNC_WRITE` as `type`. **]**
+
+**SRS_FILE_LINUX_43_020: [** `file_write_async` shall populate the `data` field of the `FILE_LINUX_IO` struct using `source`, `user_callback` and `user_context`. **]**
+
+**SRS_FILE_LINUX_43_004: [** `file_write_async` shall initialize a `sigevent` struct with `SIGEV_THREAD` as `sigev_notify`, the created `FILE_LINUX_IO` struct as `sigev_value`, `on_file_io_complete_linux` as `sigev_notify_function`, `NULL` as `sigev_notify_attributes`(unsure about what `sigev_notify_attributes` should be). **]**
+
+**SRS_FILE_LINUX_43_005: [** `file_write_async` shall initialize an `aiocb` struct with `handle` as `aio_fildes`, `position` as `aio_offset`, source as `aio_buf`, `source->alias->size` as `aio_nbytes`, and the initialized `sigevent` struct as `aio_sigevent`. **]**
+
+**SRS_FILE_LINUX_43_006: [** `file_write_async` shall call `aio_write` with the initialized `aiocb` struct as `aiocbp`.**]**
+
+**SRS_FILE_LINUX_43_012: [** If `aio_write` fails, `file_write_async` shall return `FILE_WRITE_ASYNC_WRITE_ERROR`. **]**
+
+**SRS_FILE_LINUX_43_007: [** If `aio_write` succeeds, `file_write_async` shall return `FILE_WRITE_ASYNC_OK`. **]**
+
+**SRS_FILE_LINUX_43_013: [** If there are any other failures, `file_write_async` shall return `FILE_WRITE_ASYNC_ERROR`. **]**
+
+## file_read_async
+
+```c
+MOCKABLE_FUNCTION_WITH_RETURNS(, FILE_READ_ASYNC_RESULT, file_read_async, FILE_HANDLE, handle, uint32_t, size, uint64_t, position, FILE_READ_CB, user_callback, void*, user_context)(0, MU_FAILURE);
+```
+
+**SRS_FILE_LINUX_43_034: [** If `handle` is `NULL` then `file_read_async` shall fail and return `FILE_READ_ASYNC_INVALID_ARGS`. **]**
+
+**SRS_FILE_LINUX_43_035: [** If `user_callback` is `NULL` then `file_read_async` shall fail and return `FILE_READ_ASYNC_INVALID_ARGS`. **]**
+
+**SRS_FILE_LINUX_43_016: [** `file_read_async` shall create a `FILE_LINUX_IO` struct with `handle` as `handle`, `FILE_ASYNC_READ` as `type`. **]**
+
+**SRS_FILE_LINUX_43_017: [** `file_read_async` shall populate the `data` field of the `FILE_LINUX_IO` struct using `user_callback` and `user_context`. **]**
+
+**SRS_FILE_LINUX_43_008: [** `file_read_async` shall initialize a `sigevent` struct with `SIGEV_THREAD` as `sigev_notify`, `user_context` as `sigev_value`, `user_callback` as `sigev_notify_function`, `NULL` as `sigev_notify_attributes`(unsure about what `sigev_notify_attributes` should be). **]**
+
+**SRS_FILE_LINUX_43_009: [** `file_read_async` shall initialize an `aiocb` struct with `handle` as `aio_fildes`, `position` as `aio_offset`, source as `aio_buf`, `source->alias->size` as `aio_nbytes`, and the initialized `sigevent` struct as `aio_sigevent`. **]**

```

Comment:
>file_read_async [](start = 30, length = 15)

Interesting. So this API will allocate memory in the kernel and return it to the user space? Where's the memory for "size" bytes allocated? [I am asking because kernel rarely will allocate memory for the user, the usual paradigm is to have the user allocate some memory and AFTER the kernel will fill the memory with some data it retrieves]

Reply:
Yeah I think it makes more sense to add an argument to accept a buffer the user has allocated.

---
In reply to: [450478296](https://github.com/Azure/azure-c-pal/pull/7#discussion_r450478296) [](ancestors = 450478296)

Reply:
I think the term "user" is overloaded here. I don't think we want to have the "user" of our API allocate, but we want to have this function allocate "in user mode" (since aio_read is going to call into the kernel at some point to do the file reading). Meaning this function should allocate the buffer to hold the data to be read.

---
In reply to: [451667784](https://github.com/Azure/azure-c-pal/pull/7#discussion_r451667784) [](ancestors = 451667784,450478296)

Reply:
Will add spec to allocate buffer.


---
In reply to: [451687850](https://github.com/Azure/azure-c-pal/pull/7#discussion_r451687850) [](ancestors = 451687850,451667784,450478296)